File: backend/services/products_service.py
from repositories.products_repository import ProductsRepository
from services.base_service import BaseService


class ProductsService(BaseService):
    def __init__(self):
        super().__init__()
        self.products_repository = self._get_repository(ProductsRepository)

    def get_all_products(self, family_id:int):
        """Fetch all products."""
        return self.products_repository.get_all_products(family_id)
    
    def add_product(self, product: dict):
        """Add a new Product."""
        return self.products_repository.add_product(product)
    
    def update_product(self, product_id:int, product: dict):
        """Update Product."""
        return self.products_repository.update_product(product_id, product)
    
    def delete_product(self, id: int):
        """Delete Product."""
        return self.products_repository.delete_product(id)

File: frontend/src/app/services/products.service.ts
import { Injectable } from '@angular/core';
import { ServiceBase } from '../shared/service.base';
import { Message, Product } from '../shared/interfaces';
import { Observable } from 'rxjs';
import { HttpClient } from '@angular/common/http';

@Injectable({
  providedIn: 'root',
})
export class ProductsService extends ServiceBase {
  constructor(private httpClient: HttpClient) {
    super();
  }
  addProduct(product: Product): Observable<Message<Product>> {
    const familyId = this.getUser()?.familyId;
    const url = this.buildUrl('products', { familyId });
    return this.pipeError(this.httpClient.post(url, product, this.httpOptions));
  }
  updateProduct(product: Product): Observable<Message<Product>> {
    const url = this.buildUrl(`products/${product.id}`);
    return this.pipeError(this.httpClient.put(url, product, this.httpOptions));
  }
  deleteProduct(productId: number): Observable<Message<Product[]>> {
    const url = this.buildUrl(`products/${productId}`);
    return this.pipeError(this.httpClient.delete(url, this.httpOptions));
  }
}
